오디오 정지, 구간 건너뛰기 기능 구현

pieisland opened the pull request:
## 📕 제목

PR 제목
관련이슈 #23 #25 

## 📗 작업 내용

> 구현 내용 및 작업 했던 내역

- [x] 정지
- [x] 앞으로 3초 건너뛰기
- [x] 뒤로 3초 건너뛰기

![ezgif com-gif-maker](https://user-images.githubusercontent.com/35261724/101234022-74426200-36ff-11eb-8102-38052da0c27d.gif)


## 📘 PR 특이 사항

> PR을 볼 때 주의깊게 봐야하거나 말하고 싶은 점

- audioContext.currentTime을 이용해 markerTime을 계산하던 방식에서 markerTime도 setInteval 내에서 주기적으로 변경하는 방식으로 바꾸었습니다.
- setInterval에 의해 stop을 눌렀을 때 바로 정지가 되는 게 아니라 bool값을 확인하는 시간 간격이 있어 초기화 부분에도 setTimeout을 사용했습니다.
- playbackTools에서 다른 곳으로 옮기려 했으나 오류도 나고 적절한 곳을 모르겠어서 옮기지 않았습니다.
- 최대한 있는 함수를 활용해 기능을 구현하려 했기 때문에 controller의 함수의 이름과 기능이 맞지 않는 부분이 있습니다. 정은님께서 이름을 수정할 예정이라고 하셨기 때문에 고치지 않았고, 다시 한 번 이름 변경을 해야 할 것으로 예상됩니다.
- 앞으로 3초 건너뛰기 시에 marker위치와 시간, playtime의 시간이 음수가 되는 경우에 대한 처리를 했습니다.
  - playtime은 시간 전체를 다시 계산해 나누는 방식을 사용하여 기존 시간을 업데이트 하는 로직과 좀 다르게 됐습니다.

pieisland committed:
feat : [FE] Audio Stop 구현

- 정지 버튼을 누르면 오디오 재생이 정지됨
- 마커의 위치가 0으로 이동
- 재생시간이 0으로 초기화
- controller에 setMarkerWidthToZero 메서드 추가
- store에 setMarkerWidthToZero 메서드 추가
- storetype으로 current_position_zero_channel 추가
- Marker는 해당 채널을 subscribe해 마커 위치를 0으로 초기화함

pieisland committed:
chore : [FE] 공백 삭제

- enter가 두 번 들어간 부분 삭제

pieisland committed:
feat : [FE] 건너뛴 시간이 음수일 때 처리

- markerTime이 음수가 됐을 때 0으로 변경
- playTime이 음수가 넘어왔을 때 재계산

pieisland committed:
feat : [FE] 마커의 위치가 음수일 때 처리

- 마커가 움직인 위치가 음수가 넘어와 현재 위치와 계산 시 음수가 됐을 때
  0으로 값 적용

pieisland committed:
feat : [FE] 구간 건너뛰기 기능 구현

- 구간 건너뛰기 기능 구현
- quantum은 3으로 설정
- 앞으로 건너뛰기, 뒤로 건너뛰기 가능
- markerTime 변경을 audioContext.currentTime을 사용하는 것이 아닌
  setInterval에서 주기적으로 변하게 하는 것으로 변경
- setInterval이 건너뛰기에서도 필요하여 따로 함수로 분리해 사용

pieisland committed:
chore : [FE] 주석 삭제

- 주석을 몇 개 삭제했습니다.
- 이해가 헷갈리는 부분들의 주석은 아직 삭제하지 않았습니다.

wooojini reviewed (approved):
고생하셨습니다!! 이제 재생기능이 점점 완성되어가네요!!! 별것 아니지만 주석이 너무 많다고 생각이 듭니다ㅠ  주석보다는 코드 자체로 이해할 수 있게 네이밍을 서술형으로 자세히 해주는 것을 어떨까요??

Songwonseok commented on frontend/src/store/store.ts:
```diff
@@ -180,7 +180,6 @@ const store = new (class Store {
     storeChannel.publish(StoreChannelType.CURRENT_POSITION_ZERO_CHANNEL, '');
   }
 
-
```
이런 커밋인 경우는 커밋 타입을 style로 해줘도 될 것 같습니당
`style   : 스타일 (코드 형식, 세미콜론 추가: 비즈니스 로직에 변경 없음)`

Jeongeun-Choi commented on frontend/src/components/PlaybackTools/PlaybackTools.ts:
```diff
@@ -147,10 +149,35 @@ const TIMER_TIME = 34;
       if (this.trackList.length == 0) return;
+
+      //일지성지 상태에서 건너뛰기 버튼을 눌렀을 때
+      if(this.isPause === true){        
```
this.isPause 자체가 true면 if문 안의 내용이 실행될거니까 this.isPause === true 대신 this.isPause가 더 간략하게 나타낼 수 있지 않을까 생각합니다!😊

pieisland committed:
feat : [FE] 맨 앞으로 건너뛰기

- 기존의 stop, play 기능을 호출하여 맨 앞으로 건너뛸 수 있도록 구현
- stop을 했을 때 delay가 발생하기 때문에 setTimeout 내에서 play를
  호출하도록 함

Jeongeun-Choi commented on frontend/src/components/PlaybackTools/PlaybackTools.ts:
```diff
@@ -200,15 +227,30 @@ const TIMER_TIME = 34;
     }
 
-    play(markerTime: number): void {
+    playTimer() {
+      let playTimer = setInterval(() => {
+        if (Controller.getIsPauseState()) {
+          clearInterval(playTimer);
+        }
+        const widthPixel = PlayBarUtil.getSomePixel(TIMER_TIME);
+        Controller.setMarkerWidth(widthPixel);
+        Controller.changePlayTime(TIMER_TIME);
+        Controller.pauseChangeMarkerTime(TIMER_TIME/1000);
+      }, TIMER_TIME);
+    }
```
play 함수 안에 있던 내용을 분리해주신거 너무 좋아요!!😁👍

Songwonseok reviewed (approved):
고생하셨어요!ㅎㅎ 
정지, 건너뛰기 기능 App컴포넌트 단축키 이벤트에 추가해주시면 좋을 것 같아요

Jeongeun-Choi commented on frontend/src/components/PlaybackTools/PlaybackTools.ts:
```diff
@@ -230,32 +271,156 @@ const TIMER_TIME = 34;
+      }, TIMER_TIME+1);
+    }
+
+    fastRewind() {
+      let markerTime = Controller.getMarkerTime();
+
+      this.stopAudioSources();
+      this.sourceInfo = [];
+
+      const widthPixel = PlayBarUtil.getSomePixel(QUANTUM * 1000);
+      Controller.setMarkerWidth(-widthPixel); //마커 위치 변경
+      Controller.pauseChangeMarkerTime(-QUANTUM); //마커 시간 변경
+      Controller.changePlayTime(-QUANTUM * 1000); //playTime(왼쪽상단 시간 부분)변경
+
+      this.trackList.forEach((track: Track) => {
```
뒤의 fastForward 함수와 구조가 비슷한데 구조를 분리할 수 있지 않을까 생각합니다! QUANTUM이 음수냐 아니냐의 차이 같아서 조건 같은걸 추가해서 예를 들어 type이 rewind일 경우에는 QUANTUM을 음수로 처리한다든지 해서 공통되는 로직을 분리하는게 좋을 것 같습니다!😊

Jeongeun-Choi reviewed (approved):
재생 구현 힘드셨을텐데 너무 잘해주셨습니다! 수고하셨어용~~😀

pieisland committed:
feat : [FE] marker position 클릭으로 변경 시 재생 위한 위치 선정

- MarkerEventUtil의 clickMarkerListener에서 play 함수를 호충하면 됨.
- 아직 play 함수를 다른 곳에서 호출할 수 없어서 위치를 잡아두기만 함.

pieisland committed:
fix : [FE] FastRewind 시의 시작 오류 수정

- fastRewind가 0이하로 됐을 때 waitTime 시간이 제대로 계산이 안되어
  delay가 생김
- markerTime - QUANTUM이 음수가 됐을 때 waitTime을 0으로 설정.

pieisland committed:
refactor : [FE] 변수 이름 변경과 주석 삭제

- when, offset, duration 변수 이름을 waitTime, audioStartTime,
  playDuration으로 변경
- 주석 삭제
